[ao] Fixing obs insertion through dtype propagation

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73006

Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6 and related to https://github.com/pytorch/pytorch/issues/72698,
when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

HDCharles committed:
[ao] Fixing obs insertion through dtype propagation

Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6
when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34302245).

vkuzo commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -455,6 +456,62 @@ def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module
+
+NON_TENSOR_ARG_DICT = {
+    NodeInfo("call_method", "masked_fill") : {
+        torch.bool: [1],
```
this is a tensor arg with dtype boolean, should we rename the data structure to communicate that it can contain indices of both tensors and non tensors?

HDCharles replied:
good idea

vkuzo commented:
I commented on the original issue with one more possible approach to solving this, thoughts on that one?

HDCharles commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -455,6 +456,62 @@ def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module
+# so that they can be propagated correctly since inserting observers
+# for them would cause errors
+
+NON_TENSOR_ARG_DICT = {
```
ok, we can do that

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6 and related to https://github.com/pytorch/pytorch/issues/72698,
when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34302245](https://our.internmc.facebook.com/intern/diff/D34302245)

[ghstack-poisoned]

vkuzo reviewed (approved):
thanks for fixing this!

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6 and related to https://github.com/pytorch/pytorch/issues/72698,
when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34302245](https://our.internmc.facebook.com/intern/diff/D34302245)

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6 and related to https://github.com/pytorch/pytorch/issues/72698,
when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[ao] Fixing obs insertion through dtype propagation"


Summary: As noticed in https://discuss.pytorch.org/t/calibration-of-model-in-post-training-static-quantization-using-fx-api/143661/6 and related to https://github.com/pytorch/pytorch/issues/72698,
when using fx quantizaiton, if an op like view was used in a
model and the index parameters were passed in to the ops with a
variable rather than
hard coded, fx would mistakenly insert observers for them, leading to an
error when the observer tried to do tensor only operations on a
non-tensor. To fix this, an API was added to specify non tensor
arguments for various ops to enable better dtype propagation.
NON_TENSOR_ARG_DICT is a nested dict whose first key is a named tuple
which contains matching parameters for ops with nontensor args, the
inner dict's keys are dtypes and the values are a list of those arg indices that
take use such dtypes. Alternatively, instead of a list, the inner dict
value can also be a function that takes the node as an argument and
returns the list of arg indices.

Theoretically this api can support arbitrary functions but the current
implmentation is limited to simpler functions given the particular
issue this fixes seems to be rare.

Note: although torch.unsqueeze and torch.transpose are listed in
quantization_patterns.py, those ops appear to be untraceable by fx. I've
included tests for their cases but fixing this issue is beyond the scope
of this PR

Test Plan:

python test/test_quantization.py test_non_reference_size
...
python test/test_quantization.py test_non_reference_<op>

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -455,6 +456,70 @@ def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module
+    Returns a dict with of non float tensor types as keys and values which correspond to a
+    function to retrieve the list (which takes the node as an argument)
+    """
+    info = NodeInfo(node.op, node.target)
```
I was mimicking existing solutions for matching nodes, is there a general best practice to define when its best to create a data structure to organize this type of data vs throwing it into a large if block?

vkuzo replied:
it was an optional comment, feel free to ignore it, up to you

> I was mimicking existing solutions for matching nodes, is there a general best practice to define when its best to create a data structure to organize this type of data vs throwing it into a large if block?

I'm a big fan of YAGNI (https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), by that principle the first use case of the code would be as simple as possible, and once there is an actual need to reuse the code would be refactored to be more reusable at that time.

On the other hand, the current PR does do it in a Pythonic way and the cognitive overhead is not too much.  Feel free to go with it.
